diff --git a/Makefile.am b/Makefile.am
index 75e376d..43e761b 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -173,7 +173,7 @@ include/event2/event-config.h: config.h make-event-config.sed
 	$(SED) -f $(srcdir)/make-event-config.sed < config.h > $@T
 	mv -f $@T $@
 
-CORE_SRC = event.c evthread.c buffer.c \
+CORE_SRC = event.c ../../mempool/s3_memory_pool.c evthread.c buffer.c \
 	bufferevent.c bufferevent_sock.c bufferevent_filter.c \
 	bufferevent_pair.c listener.c bufferevent_ratelim.c \
 	evmap.c	log.c evutil.c evutil_rand.c strlcpy.c $(SYS_SRC)
@@ -213,7 +213,7 @@ libevent_openssl_la_LIBADD = $(MAYBE_CORE) $(OPENSSL_LIBS)
 libevent_openssl_la_LDFLAGS = $(GENERIC_LDFLAGS)
 endif
 
-noinst_HEADERS = util-internal.h mm-internal.h ipv6-internal.h \
+noinst_HEADERS = util-internal.h ../../mempool/s3_memory_pool.h mm-internal.h ipv6-internal.h \
 	evrpc-internal.h strlcpy-internal.h evbuffer-internal.h \
 	bufferevent-internal.h http-internal.h event-internal.h \
 	evthread-internal.h ht-internal.h defer-internal.h \
@@ -232,7 +232,7 @@ else
 noinst_HEADERS += $(EVENT1_HDRS)
 endif
 
-AM_CPPFLAGS = -I$(srcdir)/compat -I$(srcdir)/include -I./include $(SYS_INCLUDES)
+AM_CPPFLAGS = -I$(srcdir)/compat -I$(srcdir)/include -I./include -I../../mempool $(SYS_INCLUDES)
 
 verify: check
 
diff --git a/buffer.c b/buffer.c
index 8510955..2ebe8be 100644
--- a/buffer.c
+++ b/buffer.c
@@ -135,6 +135,7 @@ static int use_mmap = 1;
 #define CHAIN_PINNED(ch)  (((ch)->flags & EVBUFFER_MEM_PINNED_ANY) != 0)
 #define CHAIN_PINNED_R(ch)  (((ch)->flags & EVBUFFER_MEM_PINNED_R) != 0)
 
+extern size_t g_event_max_read;
 static void evbuffer_chain_align(struct evbuffer_chain *chain);
 static int evbuffer_chain_should_realign(struct evbuffer_chain *chain,
     size_t datalen);
@@ -156,12 +157,10 @@ evbuffer_chain_new(size_t size)
 {
 	struct evbuffer_chain *chain;
 	size_t to_alloc;
-
-	if (size > EVBUFFER_CHAIN_MAX - EVBUFFER_CHAIN_SIZE)
+	size_t pool_item_size = 0;
+	if (size > EVBUFFER_CHAIN_MAX)
 		return (NULL);
 
-	size += EVBUFFER_CHAIN_SIZE;
-
 	/* get the next largest memory that can hold the buffer */
 	if (size < EVBUFFER_CHAIN_MAX / 2) {
 		to_alloc = MIN_BUFFER_SIZE;
@@ -172,18 +171,32 @@ evbuffer_chain_new(size_t size)
 		to_alloc = size;
 	}
 
-	/* we get everything in one chunk */
-	if ((chain = mm_malloc(to_alloc)) == NULL)
+	/* Allocating the structure, buffer within this structure will be
+	 * allocated separately */
+	if ((chain = mm_malloc(EVBUFFER_CHAIN_SIZE)) == NULL)
 		return (NULL);
 
-	memset(chain, 0, EVBUFFER_CHAIN_SIZE);
-
-	chain->buffer_len = to_alloc - EVBUFFER_CHAIN_SIZE;
+	mempool_getbuffer_size(g_pool_handle, &pool_item_size);
 
-	/* this way we can manipulate the buffer to different addresses,
-	 * which is required for mmap for example.
-	 */
-	chain->buffer = EVBUFFER_CHAIN_EXTRA(u_char, chain);
+	memset(chain, 0, EVBUFFER_CHAIN_SIZE);
+	if ((pool_item_size >= to_alloc) && (pool_item_size % to_alloc == 0)) {
+		if ((chain->buffer = (unsigned char *)mempool_getbuffer(
+			g_pool_handle, ZEROED_ALLOCATION)) == NULL) {
+			mm_free(chain);
+			return NULL;
+		}
+		chain->buffer_len = pool_item_size;
+		chain->chain_buffer_len = pool_item_size;
+	} else {
+		if ((chain->buffer = (unsigned char *)mm_malloc(to_alloc)) ==
+		    NULL) {
+			mm_free(chain);
+			return NULL;
+		}
+		chain->buffer_len = to_alloc;
+		chain->chain_buffer_len = to_alloc;
+	}
+	chain->chain_buffer = chain->buffer;
 
 	return (chain);
 }
@@ -191,6 +204,7 @@ evbuffer_chain_new(size_t size)
 static inline void
 evbuffer_chain_free(struct evbuffer_chain *chain)
 {
+	size_t pool_item_size = 0;
 	if (CHAIN_PINNED(chain)) {
 		chain->flags |= EVBUFFER_DANGLING;
 		return;
@@ -231,6 +245,15 @@ evbuffer_chain_free(struct evbuffer_chain *chain)
 #endif
 	}
 
+	mempool_getbuffer_size(g_pool_handle, &pool_item_size);
+
+	if (chain->chain_buffer_len == pool_item_size) {
+		/* If the buffer size is same as that of pool's buffer size then
+		 * this have been allocated from pool, so release to pool */
+		mempool_releasebuffer(g_pool_handle, chain->chain_buffer);
+	} else {
+		mm_free(chain->chain_buffer);
+	}
 	mm_free(chain);
 }
 
@@ -640,8 +663,13 @@ evbuffer_reserve_space(struct evbuffer *buf, ev_ssize_t size,
 	} else {
 		if (_evbuffer_expand_fast(buf, size, n_vecs)<0)
 			goto done;
+		/* last argument set to 1, to allocate the 'size' space
+      	 * in the vectors, even if more space is available.
+		 */
+		/* S3 edit starts */
 		n = _evbuffer_read_setup_vecs(buf, size, vec, n_vecs,
-				&chainp, 0);
+				&chainp, 1);
+		/* S3 edit ends */
 	}
 
 done:
@@ -2013,7 +2041,7 @@ evbuffer_expand(struct evbuffer *buf, size_t datlen)
 #endif
 #define NUM_READ_IOVEC 4
 
-#define EVBUFFER_MAX_READ	4096
+#define EVBUFFER_MAX_READ	16384
 
 /** Helper function to figure out which space to use for reading data into
     an evbuffer.  Internal use only.
@@ -2074,12 +2102,23 @@ get_n_bytes_readable_on_socket(evutil_socket_t fd)
 	/* Can overflow, but mostly harmlessly. XXXX */
 	return (int)lng;
 #elif defined(FIONREAD)
-	int n = EVBUFFER_MAX_READ;
+	int n;
+	if (g_event_max_read == 0)
+	{
+		n = EVBUFFER_MAX_READ;
+	} else {
+		n = g_event_max_read;
+	}
 	if (ioctl(fd, FIONREAD, &n) < 0)
 		return -1;
 	return n;
 #else
-	return EVBUFFER_MAX_READ;
+	if (g_event_max_read == 0)
+	{
+		return EVBUFFER_MAX_READ;
+	} else {
+		return g_event_max_read;
+	}
 #endif
 }
 
@@ -2107,11 +2146,14 @@ evbuffer_read(struct evbuffer *buf, evutil_socket_t fd, int howmuch)
 	}
 
 	n = get_n_bytes_readable_on_socket(fd);
-	if (n <= 0 || n > EVBUFFER_MAX_READ)
-		n = EVBUFFER_MAX_READ;
+	if (g_event_max_read == 0)
+		g_event_max_read = EVBUFFER_MAX_READ;
+
+	if (n <= 0 || n > g_event_max_read)
+		n = g_event_max_read;
+
 	if (howmuch < 0 || howmuch > n)
 		howmuch = n;
-
 #ifdef USE_IOVEC_IMPL
 	/* Since we can use iovecs, we're willing to use the last
 	 * NUM_READ_IOVEC chains. */
@@ -2762,6 +2804,7 @@ evbuffer_add_reference(struct evbuffer *outbuf,
 		/* don't call chain_free; we do not want to actually invoke
 		 * the cleanup function */
 		mm_free(chain);
+		mm_free(info);
 		goto done;
 	}
 	evbuffer_chain_insert(outbuf, chain);
@@ -2827,6 +2870,7 @@ evbuffer_add_file(struct evbuffer *outbuf, int fd,
 		EVBUFFER_LOCK(outbuf);
 		if (outbuf->freeze_end) {
 			mm_free(chain);
+			mm_free(info);
 			ok = 0;
 		} else {
 			outbuf->n_add_for_cb += length;
diff --git a/bufferevent.c b/bufferevent.c
index 7518c3b..6dc18d0 100644
--- a/bufferevent.c
+++ b/bufferevent.c
@@ -510,9 +510,19 @@ bufferevent_disable(struct bufferevent *bufev, short event)
 }
 
 /*
- * Sets the water marks
+ * Set the maximum read
  */
+void
+bufferevent_set_read_data(struct bufferevent *bufev, size_t max_val)
+{
+	BEV_LOCK(bufev);
+	bufev->read_max_data = max_val;
+	BEV_UNLOCK(bufev);
+}
 
+/*
+ * Sets the water marks
+ */
 void
 bufferevent_setwatermark(struct bufferevent *bufev, short events,
     size_t lowmark, size_t highmark)
diff --git a/bufferevent_openssl.c b/bufferevent_openssl.c
index 7582d9b..16881ed 100644
--- a/bufferevent_openssl.c
+++ b/bufferevent_openssl.c
@@ -61,6 +61,8 @@
 #include <openssl/ssl.h>
 #include <openssl/err.h>
 
+#define EVBUFFER_MAX_READ	16384
+
 /*
  * Define an OpenSSL bio that targets a bufferevent.
  */
@@ -99,6 +101,10 @@ print_err(int val)
 #define print_err(v) ((void)0)
 #endif
 
+/* S3 edit starts */
+extern size_t g_event_max_read;
+/* S3 edit ends */
+
 /* Called to initialize a new BIO */
 static int
 bio_bufferevent_new(BIO *b)
@@ -321,6 +327,9 @@ struct bufferevent_openssl {
 
 	/* Are we currently connecting, accepting, or doing IO? */
 	unsigned state : 2;
+	/* S3 edit starts */
+	size_t read_max_data;
+	/* S3 edit ends */
 };
 
 static int be_openssl_enable(struct bufferevent *, short);
@@ -383,6 +392,43 @@ start_reading(struct bufferevent_openssl *bev_ssl)
 		struct bufferevent *bev = &bev_ssl->bev.bev;
 		int r;
 		r = _bufferevent_add_event(&bev->ev_read, &bev->timeout_read);
+		/* S3 edit starts */
+		/*
+		 * This is a fix for EOS-198. Server would crash or return
+		 * timeout errors, when is tested with cosbench with SSL
+		 * enabled.  Specific scenario ivolved writing objects of size
+		 * 1000000 bytes.  Current settings force libevent to read data
+		 * in chunks of 16k, by setting low- and high-water marks to
+		 * 16k.  When reading in 16k chunks, the last chunk of data is
+		 * 576 bytes.  The issue that was observed was: libevent times
+		 * out on read, as if those 576 bytes were not sent by client.
+		 * When timeout has triggered, and S3 server has already sent
+		 * error response to client, right after that, libevent
+		 * suddenly discoveres that there's data for reading, and calls
+		 * consider_reading.  Crash was caused probably because
+		 * consider_reading calls htparser_run, and it uses parser
+		 * struct that has been already reset.
+		 *
+		 * The fix is -- if at the end of payload we see that there's
+		 * data pending, then immediately activate the event -- don't
+		 * wait until new data comes to socket.
+		 *
+		 * Since this is risky fix, condition for event activation is
+		 * specified as narrow as possible.  We check that watermark
+		 * (wm_read.low) is reduced -- this happens right after we read
+		 * last full chunk of 16k.  We check that pending data is more
+		 * or equal to watermark -- if it's less than watermark, it
+		 * means there's more data in socket, and we don't need to act.
+		 * We check that pending is more than 0 -- if it's 0, there
+		 * must be more data in socket.
+		 */
+		if (bev->wm_read.low < g_event_max_read &&
+		    SSL_pending(bev_ssl->ssl) >= bev->wm_read.low &&
+		    SSL_pending(bev_ssl->ssl) > 0) {
+			evmap_io_active(bev->ev_base, bev->ev_read.ev_fd,
+					EV_READ | EV_ET);
+		}
+		/* S3 edit ends */
 		if (r == 0 && bev_ssl->read_blocked_on_write)
 			r = _bufferevent_add_event(&bev->ev_write,
 			    &bev->timeout_write);
@@ -748,6 +794,16 @@ bytes_to_read(struct bufferevent_openssl *bev)
 		result = READ_DEFAULT;
 	}
 
+	/* read only 'read_max_data' bytes, when `read_max_data'
+	 * value is '0'. read_max_data will set only for the first time
+	 * and is to read the bytes which is equal to header size.
+	 */
+	/* S3 edit starts */
+	if (bev->read_max_data != 0) {
+		result = bev->read_max_data;
+	}
+	/* S3 edit ends */
+
 	/* Respect the rate limit */
 	limit = _bufferevent_get_read_max(&bev->bev);
 	if (result > limit) {
@@ -778,6 +834,10 @@ consider_reading(struct bufferevent_openssl *bev_ssl)
 		return;
 
 	n_to_read = bytes_to_read(bev_ssl);
+	// reset to '0'
+	/* S3 edit starts */
+	bev_ssl->read_max_data = 0;
+	/* S3 edit ends */
 
 	while (n_to_read) {
 		r = do_read(bev_ssl, n_to_read);
@@ -799,6 +859,19 @@ consider_reading(struct bufferevent_openssl *bev_ssl)
 		 * whole SSL record anyway. */
 		n_to_read = SSL_pending(bev_ssl->ssl);
 
+		/* S3 edit starts */
+		/* Read only the number of bytes required to make evbuffer
+		 * size alignment with 16k, instead of reading all
+		 * available bytes retunred by SSL_pending call.
+		 */
+		if (n_to_read > 0) {
+			struct bufferevent *bev = &bev_ssl->bev.bev;
+			struct evbuffer *input = bev->input;
+			int bytes_pending_to_read = EVBUFFER_MAX_READ - evbuffer_get_length(input);
+			n_to_read = bytes_pending_to_read > 0 ? bytes_pending_to_read : 0;
+		}
+		/* S3 edit ends */
+
 		/* XXX This if statement is actually a bad bug, added to avoid
 		 * XXX a worse bug.
 		 *
@@ -1452,3 +1525,14 @@ bufferevent_get_openssl_error(struct bufferevent *bev)
 	BEV_UNLOCK(bev);
 	return err;
 }
+
+/* S3 edit starts */
+void bufferevent_ssl_set_read_data(struct bufferevent *bev, size_t max_val)
+{
+    struct bufferevent_openssl *bev_ssl;
+    BEV_LOCK(bev);
+    bev_ssl = upcast(bev);
+    bev_ssl->read_max_data = max_val;
+    BEV_UNLOCK(bev);
+}
+/* S3 edit ends */
diff --git a/bufferevent_sock.c b/bufferevent_sock.c
index 9b0caf2..5ceb1f3 100644
--- a/bufferevent_sock.c
+++ b/bufferevent_sock.c
@@ -154,6 +154,10 @@ bufferevent_readcb(evutil_socket_t fd, short event, void *arg)
 			goto done;
 		}
 	}
+
+	if (bufev->read_max_data != 0) {
+		howmuch = bufev->read_max_data;
+	}
 	readmax = _bufferevent_get_read_max(bufev_p);
 	if (howmuch < 0 || howmuch > readmax) /* The use of -1 for "unlimited"
 					       * uglifies this code. XXXX */
@@ -179,6 +183,8 @@ bufferevent_readcb(evutil_socket_t fd, short event, void *arg)
 	if (res <= 0)
 		goto error;
 
+	bufev->read_max_data = 0;
+
 	_bufferevent_decrement_read_buckets(bufev_p, res);
 
 	/* Invoke the user callback - must always be called last */
diff --git a/evbuffer-internal.h b/evbuffer-internal.h
index f9dcc06..c105fcc 100644
--- a/evbuffer-internal.h
+++ b/evbuffer-internal.h
@@ -52,7 +52,7 @@ extern "C" {
 #if _EVENT_SIZEOF_VOID_P < 8
 #define MIN_BUFFER_SIZE	512
 #else
-#define MIN_BUFFER_SIZE	1024
+#define MIN_BUFFER_SIZE	4096
 #endif
 
 /** A single evbuffer callback for an evbuffer. This function will be invoked
@@ -173,6 +173,9 @@ struct evbuffer_chain {
 	/** total allocation available in the buffer field. */
 	size_t buffer_len;
 
+	/** Total allocation created initially */
+	size_t chain_buffer_len;
+
 	/** unused space at the beginning of buffer or an offset into a
 	 * file for sendfile buffers. */
 	ev_misalign_t misalign;
@@ -204,6 +207,9 @@ struct evbuffer_chain {
 	 * may point to NULL.
 	 */
 	unsigned char *buffer;
+	/* Will hold the initial address of buffer, will be used to free the
+	 * same */
+	unsigned char *chain_buffer;
 };
 
 /* this is currently used by both mmap and sendfile */
@@ -223,7 +229,8 @@ struct evbuffer_chain_reference {
 
 #define EVBUFFER_CHAIN_SIZE sizeof(struct evbuffer_chain)
 /** Return a pointer to extra data allocated along with an evbuffer. */
-#define EVBUFFER_CHAIN_EXTRA(t, c) (t *)((struct evbuffer_chain *)(c) + 1)
+#define EVBUFFER_CHAIN_EXTRA(t, c)                                             \
+	(t *)(((struct evbuffer_chain *)(c))->chain_buffer)
 
 /** Assert that we are holding the lock on an evbuffer */
 #define ASSERT_EVBUFFER_LOCKED(buffer)			\
diff --git a/event.c b/event.c
index fab419a..e46e9a0 100644
--- a/event.c
+++ b/event.c
@@ -25,6 +25,7 @@
  * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 #include "event2/event-config.h"
+#include "s3_memory_pool.h"
 
 #ifdef WIN32
 #include <winsock2.h>
@@ -125,6 +126,10 @@ struct event_base *event_global_current_base_ = NULL;
 
 static int use_monotonic;
 
+/* global Memory Pool handle */
+MemoryPoolHandle g_pool_handle;
+/* global max allowed read */
+size_t g_event_max_read;
 /* Prototypes */
 static inline int event_add_internal(struct event *ev,
     const struct timeval *tv, int tv_is_absolute);
@@ -2726,6 +2731,35 @@ event_set_mem_functions(void *(*malloc_fn)(size_t sz),
 	_mm_free_fn = free_fn;
 }
 #endif
+int
+event_use_mempool(size_t item_size, size_t pool_initial_size,
+		     size_t pool_expansion_size,
+		     size_t pool_max_threshold_size,
+		     int flags)
+{
+	int rc;
+	rc = mempool_create(item_size, pool_initial_size, pool_expansion_size,
+			pool_max_threshold_size, flags, &g_pool_handle);
+	if (rc != 0)
+		return rc;
+
+	return 0;
+}
+
+int event_mempool_free_space(size_t *free_space) {
+	return mempool_free_space(g_pool_handle, free_space);
+}
+
+int event_destroy_mempool() {
+	return mempool_destroy(&g_pool_handle);
+
+}
+
+void event_set_max_read(size_t max_read)
+{
+	g_event_max_read = max_read;
+}
+
 
 #if defined(_EVENT_HAVE_EVENTFD) && defined(_EVENT_HAVE_SYS_EVENTFD_H)
 static void
diff --git a/include/event2/buffer.h b/include/event2/buffer.h
index 5c89655..5852e4c 100644
--- a/include/event2/buffer.h
+++ b/include/event2/buffer.h
@@ -834,6 +834,10 @@ struct event_base;
    of execution.
  */
 int evbuffer_defer_callbacks(struct evbuffer *buffer, struct event_base *base);
+/**
+  Set the maximum allowed read
+*/
+void evbuffer_set_max_read(size_t max_read);
 
 #ifdef __cplusplus
 }
diff --git a/include/event2/bufferevent.h b/include/event2/bufferevent.h
index 0cc6cb1..901f55c 100644
--- a/include/event2/bufferevent.h
+++ b/include/event2/bufferevent.h
@@ -455,6 +455,16 @@ int bufferevent_set_timeouts(struct bufferevent *bufev,
     const struct timeval *timeout_read, const struct timeval *timeout_write);
 
 /**
+  Gets the low water mark for the buffer set during memory pool creation.
+*/
+size_t  bufferevent_get_s3_low_watermark();
+
+/**
+  Sets the data length to be read from socket
+*/
+void bufferevent_set_read_data(struct bufferevent *bufev, size_t max_val);
+
+/**
   Sets the watermarks for read and write events.
 
   On input, a bufferevent does not invoke the user read callback unless
diff --git a/include/event2/bufferevent_ssl.h b/include/event2/bufferevent_ssl.h
index c612e99..20e7427 100644
--- a/include/event2/bufferevent_ssl.h
+++ b/include/event2/bufferevent_ssl.h
@@ -98,6 +98,11 @@ int bufferevent_ssl_renegotiate(struct bufferevent *bev);
 /** Return the most recent OpenSSL error reported on an SSL bufferevent. */
 unsigned long bufferevent_get_openssl_error(struct bufferevent *bev);
 
+/* S3 edit starts */
+void bufferevent_ssl_set_read_data(struct bufferevent *bev, size_t max_val);
+/* S3 edit ends */
+
+
 #endif
 
 #ifdef __cplusplus
diff --git a/include/event2/bufferevent_struct.h b/include/event2/bufferevent_struct.h
index 02835c1..fa75f51 100644
--- a/include/event2/bufferevent_struct.h
+++ b/include/event2/bufferevent_struct.h
@@ -107,6 +107,7 @@ struct bufferevent {
 	/** Events that are currently enabled: currently EV_READ and EV_WRITE
 	    are supported. */
 	short enabled;
+	size_t read_max_data;
 };
 
 #ifdef __cplusplus
diff --git a/include/event2/event.h b/include/event2/event.h
index 10bcd97..c8bf810 100644
--- a/include/event2/event.h
+++ b/include/event2/event.h
@@ -295,6 +295,41 @@ struct event_config
 ;
 
 /**
+ * Enable use of memory pool for libevent.
+ * Memory pool is created with some pre allocated buffers of fixed size.
+ * Memory will be allocated/released from/to the buffer pool for the libevent
+ * buffers.
+ *
+ * @param item_size the buffer size in the pool
+ * @param pool_initial_size the pool's initial preallocated memory
+ * @param pool_expansion_size the pool's memory expansion size
+ * @param pool_max_threshold_size Maximum outstanding memory that can be alloted
+ * from pool
+ * @flags CREATE_ALIGNED_MEMORY, ENABLE_LOCKING
+ * @return 0 if successful otherwise an error
+ */
+int event_use_mempool(size_t item_size, size_t pool_initial_size,
+			size_t pool_expansion_size,
+			size_t pool_max_threshold_size, int flags);
+
+/**
+ * Free space in the memory pool
+ */
+int event_mempool_free_space(size_t *free_space);
+
+/**
+ * Destroy the memory pool
+ */
+int event_destroy_mempool();
+
+/**
+ * Set maximum read possible.
+ *
+ * @param max_read maximum read possible
+ */
+void event_set_max_read(size_t max_read);
+
+/**
  * Enable some relatively expensive debugging checks in Libevent that
  * would normally be turned off.  Generally, these checks cause code that
  * would otherwise crash mysteriously to fail earlier with an assertion
diff --git a/mm-internal.h b/mm-internal.h
index b375155..a82e9f4 100644
--- a/mm-internal.h
+++ b/mm-internal.h
@@ -27,6 +27,7 @@
 #define _EVENT_MM_INTERNAL_H
 
 #include <sys/types.h>
+#include "s3_memory_pool.h"
 
 #ifdef __cplusplus
 extern "C" {
@@ -54,6 +55,8 @@ void event_mm_free_(void *p);
 #define mm_free(p) free(p)
 #endif
 
+/* Memory Pool handle declaration */
+extern MemoryPoolHandle g_pool_handle;
 #ifdef __cplusplus
 }
 #endif
diff --git a/test/Makefile.am b/test/Makefile.am
index 0253a49..d55aea4 100644
--- a/test/Makefile.am
+++ b/test/Makefile.am
@@ -6,7 +6,7 @@
 
 AUTOMAKE_OPTIONS = foreign
 
-AM_CPPFLAGS = -I$(top_srcdir) -I$(top_srcdir)/compat -I$(top_srcdir)/include -I../include -DTINYTEST_LOCAL
+AM_CPPFLAGS = -I$(top_srcdir) -I$(top_srcdir)/compat -I$(top_srcdir)/include -I../include -I ../../../mempool -DTINYTEST_LOCAL
 
 EXTRA_DIST = regress.rpc regress.gen.h regress.gen.c rpcgen_wrapper.sh test.sh
 
